pkg/forward/forward.go

```diff
@@ -416,6 +435,91 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
+		return true
+	}
+
+	// send to UDF only the data messages
+	if len(dataMessages) != 0 {
+		for _, m := range dataMessages {
+			// emit message size metric
+			readBytesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": isdf.fromBuffer.GetName()}).Add(float64(len(m.Payload)))
+			// assign watermark to the message. assign time.UnixMilli(-1) as watermark when we are at source vertex.
+			m.Watermark = time.Time(processorWM)
+
+			// process the UDF and get the result
+			start := time.Now()
+			udfReadMessagesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": isdf.fromBuffer.GetName()}).Inc()
+
+			writeMessageCh := make(chan isb.WriteMessage)
+			errs, ctx := errgroup.WithContext(ctx)
+			errs.Go(func() error {
+				return isdf.UDF.ApplyMapStream(ctx, m, writeMessageCh)
+			})
+
+			// Stream the message to the next vertex. First figure out which vertex
+			// to send the result to. Then update the toBuffer(s) with writeMessage.
+			for writeMessage := range writeMessageCh {
+				// if we do not get a time from UDF, we set it to the time from (N-1)th vertex
+				if writeMessage.EventTime.IsZero() {
+					writeMessage.EventTime = m.EventTime
+				}
+
+				udfWriteMessagesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": isdf.fromBuffer.GetName()}).Add(float64(1))
+
+				// update toBuffers
+				if err := isdf.whereToStep(&writeMessage, messageToStep, m); err != nil {
+					isdf.opts.logger.Errorw("failed in whereToStep", zap.Error(err))
+					isdf.fromBuffer.NoAck(ctx, readOffsets)
+					return true
+				}
+			}
+
+			// look for errors in udf processing, if we see even 1 error NoAck all messages
+			// then return. Handling partial retrying is not worth ATM.
+			if err := errs.Wait(); err != nil {
+				udfError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName,
+					"buffer": isdf.fromBuffer.GetName()}).Inc()
+				isdf.opts.logger.Errorw("failed to applyUDF", zap.Error(err))
+				// We do not retry as we are streaming
+				if ok, _ := isdf.IsShuttingDown(); ok {
+					isdf.opts.logger.Errorw("UDF.Apply, Stop called while stuck on an internal error", zap.Error(err))
+					platformError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName}).Inc()
+				}
+				// As there's no partial failure, non-ack all the readOffsets
+				isdf.fromBuffer.NoAck(ctx, readOffsets)
+				return true
+			}
+
+			udfProcessingTime.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName,
+				"buffer": isdf.fromBuffer.GetName()}).Observe(float64(time.Since(start).Microseconds()))
+		}
+	}
+
+	// forward the message to the edge buffer (could be multiple edges)
+	err := isdf.writeToBuffers(ctx, messageToStep, writeOffsets)
```

Comment:
This means we only write to next buffers after receiving all the returned messages from the channel?

Reply:
Good catch! Initially I have inside the loop for processing each received messages. Then realized non data messages also need forwarding. I guess I need to add it for both cases.